File: mm-swift/Sources/MetaMessage/Ir/Mime.swift
import Foundation

enum Mime {
    private static let strToMime: [String: UInt64] = {
        var dict: [String: UInt64] = [:]
        dict["image/jpeg"] = 1
        dict["image/jpg"] = 1
        dict["image/png"] = 2
        dict["image/gif"] = 3
        dict["image/webp"] = 4
        dict["image/svg+xml"] = 5
        dict["image/avif"] = 6
        dict["image/bmp"] = 7
        dict["image/x-icon"] = 8
        dict["image/tiff"] = 9
        dict["image/heic"] = 10
        dict["image/heif"] = 11
        dict["text/plain"] = 12
        dict["text/html"] = 13
        dict["text/css"] = 14
        dict["text/javascript"] = 15
        dict["application/json"] = 16
        dict["text/csv"] = 17
        dict["text/markdown"] = 18
        dict["application/pdf"] = 19
        dict["application/zip"] = 20
        dict["application/gzip"] = 21
        dict["application/x-tar"] = 22
        dict["application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"] = 23
        dict["application/vnd.openxmlformats-officedocument.wordprocessingml.document"] = 24
        dict["application/octet-stream"] = 25
        dict["video/mp4"] = 26
        dict["video/webm"] = 27
        dict["video/quicktime"] = 28
        dict["audio/mpeg"] = 29
        dict["audio/wav"] = 30
        dict["audio/flac"] = 31
        dict["font/woff2"] = 32
        dict["font/ttf"] = 33
        return dict
    }()

    static func parse(_ s: String) -> UInt64 {
        let key = s.trimmingCharacters(in: .whitespaces).lowercased()
        return strToMime[key] ?? 0
    }

    static func toString(_ code: UInt64) -> String {
        switch code {
        case 1: return "image/jpeg"
        case 2: return "image/png"
        case 3: return "image/gif"
        case 4: return "image/webp"
        case 5: return "image/svg+xml"
        case 6: return "image/avif"
        case 7: return "image/bmp"
        case 8: return "image/x-icon"
        case 9: return "image/tiff"
        case 10: return "image/heic"
        case 11: return "image/heif"
        case 12: return "text/plain"
        case 13: return "text/html"
        case 14: return "text/css"
        case 15: return "text/javascript"
        case 16: return "application/json"
        case 17: return "text/csv"
        case 18: return "text/markdown"
        case 19: return "application/pdf"
        case 20: return "application/zip"
        case 21: return "application/gzip"
        case 22: return "application/x-tar"
        case 23: return "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet"
        case 24: return "application/vnd.openxmlformats-officedocument.wordprocessingml.document"
        case 25: return "application/octet-stream"
        case 26: return "video/mp4"
        case 27: return "video/webm"
        case 28: return "video/quicktime"
        case 29: return "audio/mpeg"
        case 30: return "audio/wav"
        case 31: return "audio/flac"
        case 32: return "font/woff2"
        case 33: return "font/ttf"
        default: return "unknown"
        }
    }
}

File: mm-rs/src/ir/mime.rs
use std::collections::HashMap;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[allow(non_camel_case_types)]
pub enum Mime {
    Unknown = 0,

    Jpeg = 1,
    Png = 2,
    Gif = 3,
    Webp = 4,
    Svg = 5,
    Avif = 6,
    Bmp = 7,
    Ico = 8,
    Tiff = 9,
    Heic = 10,
    Heif = 11,

    TextPlain = 12,
    Html = 13,
    Css = 14,
    JavaScript = 15,
    Json = 16,
    Csv = 17,
    Markdown = 18,

    Pdf = 19,
    Zip = 20,
    Gzip = 21,
    Tar = 22,
    Xlsx = 23,
    Docx = 24,
    OctetStream = 25,

    Mp4 = 26,
    Webm = 27,
    Mov = 28,

    Mp3 = 29,
    Wav = 30,
    Flac = 31,

    Woff2 = 32,
    Ttf = 33,
}

pub fn parse_mime(s: &str) -> u8 {
    let s = s.to_lowercase();
    let s = s.trim();
    let map = mime_str_to_id();
    map.get(s).copied().unwrap_or(Mime::Unknown as u8)
}

pub fn mime_to_str(id: u8) -> &'static str {
    match id {
        1 => "image/jpeg",
        2 => "image/png",
        3 => "image/gif",
        4 => "image/webp",
        5 => "image/svg+xml",
        6 => "image/avif",
        7 => "image/bmp",
        8 => "image/x-icon",
        9 => "image/tiff",
        10 => "image/heic",
        11 => "image/heif",
        12 => "text/plain",
        13 => "text/html",
        14 => "text/css",
        15 => "text/javascript",
        16 => "application/json",
        17 => "text/csv",
        18 => "text/markdown",
        19 => "application/pdf",
        20 => "application/zip",
        21 => "application/gzip",
        22 => "application/x-tar",
        23 => "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        24 => "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        25 => "application/octet-stream",
        26 => "video/mp4",
        27 => "video/webm",
        28 => "video/mov",
        29 => "audio/mpeg",
        30 => "audio/wav",
        31 => "audio/flac",
        32 => "font/woff2",
        33 => "font/ttf",
        _ => "unknown",
    }
}

fn mime_str_to_id() -> HashMap<&'static str, u8> {
    let mut m = HashMap::new();
    m.insert("image/jpeg", Mime::Jpeg as u8);
    m.insert("image/jpg", Mime::Jpeg as u8);
    m.insert("image/png", Mime::Png as u8);
    m.insert("image/gif", Mime::Gif as u8);
    m.insert("image/webp", Mime::Webp as u8);
    m.insert("image/svg+xml", Mime::Svg as u8);
    m.insert("image/avif", Mime::Avif as u8);
    m.insert("image/bmp", Mime::Bmp as u8);
    m.insert("image/x-icon", Mime::Ico as u8);
    m.insert("image/tiff", Mime::Tiff as u8);
    m.insert("image/heic", Mime::Heic as u8);
    m.insert("image/heif", Mime::Heif as u8);
    m.insert("text/plain", Mime::TextPlain as u8);
    m.insert("text/html", Mime::Html as u8);
    m.insert("text/css", Mime::Css as u8);
    m.insert("text/javascript", Mime::JavaScript as u8);
    m.insert("application/json", Mime::Json as u8);
    m.insert("text/csv", Mime::Csv as u8);
    m.insert("text/markdown", Mime::Markdown as u8);
    m.insert("application/pdf", Mime::Pdf as u8);
    m.insert("application/zip", Mime::Zip as u8);
    m.insert("application/gzip", Mime::Gzip as u8);
    m.insert("application/x-tar", Mime::Tar as u8);
    m.insert(
        "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
        Mime::Xlsx as u8,
    );
    m.insert(
        "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
        Mime::Docx as u8,
    );
    m.insert("application/octet-stream", Mime::OctetStream as u8);
    m.insert("video/mp4", Mime::Mp4 as u8);
    m.insert("video/webm", Mime::Webm as u8);
    m.insert("video/mov", Mime::Mov as u8);
    m.insert("audio/mpeg", Mime::Mp3 as u8);
    m.insert("audio/wav", Mime::Wav as u8);
    m.insert("audio/flac", Mime::Flac as u8);
    m.insert("font/woff2", Mime::Woff2 as u8);
    m.insert("font/ttf", Mime::Ttf as u8);
    m
}